source/server/listener_manager_impl.cc

```diff
@@ -211,27 +197,29 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, ListenerManag
     ProtobufTypes::MessagePtr message =
         Config::Utility::translateToFactoryConfig(transport_socket, config_factory);
 
-    // Each transport socket factory owns one SslServerContext, we need to store them all in a
-    // vector since Ssl::ContextManager doesn't own SslServerContext. While transportSocketFacotry()
-    // always returns the first element of transport_socket_factories_, other transport socket
-    // factories are needed when the default Ssl::ServerContext updates SSL context based on
-    // ClientHello. This behavior is a workaround for initial SNI support before the full SNI based
-    // filter chain match is implemented.
-    transport_socket_factories_.emplace_back(config_factory.createTransportSocketFactory(
-        name_, sni_domains, skip_context_update, *message, *this));
-    ASSERT(transport_socket_factories_.back() != nullptr);
+    // TODO(PiotrSikora): reject multiple filter chains with the same matching criteria.
+    addFilterChain(server_names, filter_chain_match.transport_protocol(),
+                   config_factory.createTransportSocketFactory(*message, *this, server_names),
+                   parent_.factory_.createNetworkFilterFactoryList(filter_chain.filters(), *this));
+
+    install_tls_inspector = is_tls || !server_names.empty();
   }
-  ASSERT(!transport_socket_factories_.empty());
 
-  // TODO(PiotrSikora): allow filter chains with mixed use of Session Ticket Keys.
-  // This doesn't work right now, because BoringSSL uses "session context" (initial SSL_CTX that
-  // accepted connection, before SNI update) for session related stuff, including Session Ticket
-  // callback, which is going to be called iff it's set on the initial SSL_CTX, even if it's not
-  // set on the current SSL_CTX that doesn't have any Session Ticket Keys configured.
-  if (has_stk != 0 && has_stk != has_tls) {
-    throw EnvoyException(fmt::format("error adding listener '{}': filter chains with mixed use of "
-                                     "Session Ticket Keys are currently not supported",
-                                     address_->asString()));
+  // Automatically inject TLS Inspector if it wasn't configured explicitly and it's needed.
```

Comment:
I wasn't involved in the initial reviews for this feature, but I'm wondering about the tradeoff here between making this explicit as behavior the user should guarantee if they want the feature vs. implicit manipulation of the filter chain?

Reply:
Well, without TLS inspector, `sni_domains` and `transport_protocol` requirements are never going to be matched, so we can either:
1. Automatically inject TLS Inspector when `sni_domains` or `transport_protocol` are specified (this is the behavior right now).
2. Reject configuration and emit error when TLS Inspector is needed (i.e. when `sni_domains` or `transport_protocol` are specified), but isn't configured.
3. Do nothing and let user figure out why things like SNI don't work.

I've heard the "config is machine generated" argument enough times to not care anymore, so let me know which one you prefer and I'm going to change the code if necessary.

Reply:
I'm fine with either (1) or (2). Definitely not (3) please. @ggreenway @htuch opinions?

Reply:
Prefer (2), since it's least surprising. Alternatively, make the TLS inspector impossible to configure by hand and then (1) is fine as well. I would find it surprising as a user if sometimes filters appear by automatic insertion and other times I'm expected or allowed to put them there by hand (or machine generation).

Reply:
If choosing between (1) and (2) I also have a mild preference for (2) so let's just go with that with a very clear error message on what the user needs to configure.

Reply:
One thing to note about (2) is that once people start adding custom transport protocols and sniffers, we might end up with config like:

```
filter_chain_match:
- sni_domains: "example.com"
- transport_protocol: "bla"
```

which we would reject without TLS inspector, but the user don't really need it here...

In theory we could limit exception only to cases when `transport_protocol: "tls"`, but then we wouldn't reject configurations without TLS inspector that list only `sni_domains` as their criteria.

Not sure if we want to deal with this now, though...

Reply:
I think (2) is better than (1). But to allow custom transport protocols with custom listener filters, we'd need a way for either to filter to say that it provides this connection data, or a way to turn off the error, or something similar.

I think (1) is a bad idea because it's very hard to change in the future, because it could make a previously-working configuration no longer working if we remove it.

With (2), if we change/remove it, we're just being less-strict in validation, so that would be an easier change to make in the future.

Reply:
(2) done, with a check that should work for custom transport protocols as well.